Implement buffered channel for pull requests (#67)
gruntwork-io/git-xargs: README.md

1. **Set execute permissions**. For example, on Linux or Mac, you'd run:

>    ```bash

Author: ```suggestion
      ```bash
```

1. **Set execute permissions**. For example, on Linux or Mac, you'd run:

      ```bash

>    chmod u+x /usr/local/bin/git-xargs
>    ```

Author: ```suggestion
      chmod u+x /usr/local/bin/git-xargs
      ```
```

1. **Check it's working**. Run the version command to ensure everything is working properly:

>    ```bash
>    git-xargs --version
>    ```

Author: ```suggestion
      ```bash
      git-xargs --version
      ```
```

1. **Ensure you have Golang installed and working properly on your system.** [Follow the official Golang install guide](https://golang.org/doc/install) to get started.

1. **Run go get to install the latest release of git-xargs**:

>    ```bash
>    go get github.com/gruntwork-io/git-xargs
>    ```

Author: ```suggestion
     ```bash
     go get github.com/gruntwork-io/git-xargs
     ```
```

1. **Ensure you have Golang installed and working properly on your system.** [Follow the official Golang install guide](https://golang.org/doc/install) to get started.

1. **Run go get to install the latest release of git-xargs**:
   ```bash
   go get github.com/gruntwork-io/git-xargs
   ```

1. **Alternatively, use go get to install a specific release of git-xargs**:

   ```bash
   go get github.com/gruntwork-io/git-xargs@v0.0.5
   ```

### Try it out!

1. **Export a valid GitHub token**. See the guide on [Github personal access
   tokens](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token)
   for information on how to generate one. For example, on Linux or Mac, you'd run:

   ```bash
   export GITHUB_OAUTH_TOKEN=<your-secret-github-oauth-token>
   ```

1. **Provide a script or command and target some repos**. Here's a simple example of running the `touch` command in
   every repo in your GitHub organization. Follow the same pattern to start running your own scripts and commands
   against your own repos!

>    ```bash
>    git-xargs \
>      --branch-name "test-branch" \
>      --commit-message "Testing git-xargs" \
>      --github-org <enter-your-github-org-name> \
>      touch git-xargs-is-awesome.txt
>    ```

Author: ```suggestion
      ```bash
      git-xargs \
        --branch-name "test-branch" \
        --commit-message "Testing git-xargs" \
        --github-org <enter-your-github-org-name> \
        touch git-xargs-is-awesome.txt
      ```
```

1. **Ensure you have Golang installed and working properly on your system.** [Follow the official Golang install guide](https://golang.org/doc/install) to get started.

1. **Run go get to install the latest release of git-xargs**:

> 

Author: ```suggestion
```

1. **Alternatively, use go get to install a specific release of git-xargs**:

>    ```bash
>    go get github.com/gruntwork-io/git-xargs@v0.0.5
>    ```

Author: ```suggestion
     ```bash
     go get github.com/gruntwork-io/git-xargs@v0.0.5
     ```
```

   tokens](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token)
   for information on how to generate one. For example, on Linux or Mac, you'd run:

>    ```bash
>    export GITHUB_OAUTH_TOKEN=<your-secret-github-oauth-token>
>    ```

Author: ```suggestion
      ```bash
      export GITHUB_OAUTH_TOKEN=<your-secret-github-oauth-token>
      ```
```

**Specifying the pause between pull requests**
The GitHub's API [integration guidelines](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits) specify that clients should pause at least 1 second in between consecutive requests against expensive endpoints, such as the one for opening a new pull request. As a result, git-xargs defaults to pausing 1 second in between each pull request that is opened. The flag `--seconds-between-prs` allows you to modify this value, so, for example, if you were to pass `--seconds-between-prs 30`

**Guidelines and observations from testing**
In local testing, GitHub's secondary rate limits tend to kick in quickly once you're running git-xargs on more than 30 repositories concurrently. In such scenarios, passing a `--seconds-between-prs` value of 12 seconds tends to provide sufficient buffer between open pull request API calls to avoid being rate limited.

> 

Reviewer: The whole readme is 💯, well-written, easy to understand, so thank you!

Thank you for including these two sections to explain when to use the bare minimum 1 second vs 12 or 30 seconds.

Author: Thank you, Rho!

```

## Rate Limiting

git-xargs attempts to be a good citizen as regards consumption of the GitHub API. git-xargs conforms to GitHub's API [integration guidelines](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits).

In addition, git-xargs includes several features and flags to help you: 

1. run jobs without tripping GitHub's rate limits
1. recover when rate limited by automatically re-trying failed pull requests again, still while honoring the GitHub rate limits

**Distinct processing channels for expensive and non-expensive work**

> git-xargs performs all work up to issuing the API call to open a pull request in parallel - according to the concurrency limits the user might optionally specify by way of the `--max-concurrent-repos` flag.

Author: ```suggestion
git-xargs distinguishes between work that is safe to perform in parallel, such as certain git operations, and work that should be done with consideration and care, such as issuing open pull requests to GitHub's API. 

Therefore, git-xargs is able to perform all safe work as quickly as possible by leveraging goroutines, while treating the more expensive open pull request API calls separately. 

 The `--max-concurrent-repos` flag allows you to specify the upper limit of concurrent goroutines that should be allowed to carry out concurrency-safe work. 
 
Pull requests are also handled on a separate channel so that they can be buffered and retried in accordance with what rate limiting feedback git-xargs is receiving from GitHub's API.  
```

In addition, git-xargs includes several features and flags to help you: 

1. run jobs without tripping GitHub's rate limits
1. recover when rate limited by automatically re-trying failed pull requests again, still while honoring the GitHub rate limits

**Distinct processing channels for expensive and non-expensive work**
git-xargs distinguishes between work that is safe to perform in parallel, such as certain git operations, and work that should be done with consideration and care, such as issuing open pull requests to GitHub's API. 

Therefore, git-xargs is able to perform all safe work as quickly as possible by leveraging goroutines, while treating the more expensive open pull request API calls separately. 

 The `--max-concurrent-repos` flag allows you to specify the upper limit of concurrent goroutines that should be allowed to carry out concurrency-safe work. 
 
Pull requests are also handled on a separate channel so that they can be buffered and retried in accordance with what rate limiting feedback git-xargs is receiving from GitHub's API.  

> This means that git-xargs performs all the work upfront that it can as quickly as possible, and then moves onto serially process the pull request jobs that have resulted from the upfront work of cloning reposotories, making file changes, checking the git worktree, etc.

Author: ```suggestion
This means that git-xargs performs all the work upfront that it can as quickly as possible, and then moves on to serially process the pull request jobs that have resulted from the upfront work of cloning reposotories, making file changes, checking the git worktree, etc.
```

**Automatic pull request retries when rate limited**
By default, git-xargs will re-attempt opening a pull request that failed due to rate limiting. The `--max-pull-request-retries` flag allows you to specify how many times you'd like a given pull request to be re-attempted in case of failure due to rate limiting. By default, the value is `3`, meaning that if you do not pass this flag, `git-xargs` will retry all rate-limit-blocked pull requests 3 times.

**Automatic backoff when rate limiting is detected**

> When git-xargs detects that it has been rate limited by GitHub, it begins requeuing failed pull requests for retry, but with an additional, larger buffer of time in between the next attempt. This larger buffer of time is intended to allow GitHub rate limit status to return to normal. The `--seconds-to-wait-when-ratelimited` flag specifies the number of seconds in addition to wait when git-xargs detects it has been rate limited. Note that this extra buffer of time is in addition to the value specified by the `--seconds-between-prs` flag.

Author: ```suggestion
When git-xargs detects that it has been rate limited by GitHub, it begins requeuing failed pull requests for retry, but with an additional, larger buffer of time in between the next attempt. This larger buffer of time is intended to allow GitHub rate limit status to return to baseline for the git-xargs client. The `--seconds-to-wait-when-ratelimited` flag specifies the number of seconds in addition to wait when git-xargs detects it has been rate limited. Note that this extra buffer of time is in addition to the value specified by the `--seconds-between-prs` flag.
```

When it is provided by GitHub, git-xargs will instead use the values of any `Retry-After` header, or the delta of seconds between the current time and the rate limit error's reset time. When these two values are not available, git-xargs falls back to the user-specified value of `--seconds-to-wait-when-ratelimited`, if the flag was passed, or the default value, which is 60 seconds.

**Specifying the pause between pull requests**
The GitHub's API [integration guidelines](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits) specify that clients should pause at least 1 second in between consecutive requests against expensive endpoints, such as the one for opening a new pull request. As a result, git-xargs defaults to pausing 1 second in between each pull request that is opened. The flag `--seconds-between-prs` allows you to modify this value. For example, if you were to pass `--seconds-between-prs 30`, then git-xargs will sleep for half a minute between issuing pull request API calls to GitHub.

**Guidelines and observations from testing**

> In local testing, GitHub's secondary rate limits tend to kick in quickly once you're running git-xargs on more than 30 repositories concurrently. In such scenarios, passing a `--seconds-between-prs` value of 12 seconds tends to provide sufficient buffer between open pull request API calls to avoid being rate limited.

Author: ```suggestion
In local testing, the actual thresholds for when GitHub's secondary rate limits kick in can vary depending on the number of repos you're targeting and how long your script or command takes to complete. Secondary rate limits have been observed on jobs targeting as few as 10 repositories. If you are the rate limiting flags with reasonable values, your job should never be rate-limited in the first place. 

If your job is consistently being rate-limited, try incrementally increasing the value you pass with the  `--seconds-between-prs` flag. Passing a higher value will increase the overall time your job takes to complete, but it will also greatly decrease the likelihood of your job tripping GitHub's rate limits at all. Passing a lower value, or not passing the flag at all, will greatly increase the likelihood your job is rate limited by GitHub. 
```

```

## Rate Limiting

git-xargs attempts to be a good citizen as regards consumption of the GitHub API. git-xargs conforms to GitHub's API [integration guidelines](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits).

In addition, git-xargs includes several features and flags to help you: 

1. run jobs without tripping GitHub's rate limits
1. recover when rate limited by automatically re-trying failed pull requests again, still while honoring the GitHub rate limits

**Distinct processing channels for expensive and non-expensive work**

> git-xargs distinguishes between work that is safe to perform in parallel, such as certain git operations, and work that should be done with consideration and care, such as issuing open pull requests to GitHub's API. 

Author: ```suggestion

git-xargs distinguishes between work that is safe to perform in parallel, such as certain git operations, and work that should be done with consideration and care, such as issuing open pull requests to GitHub's API. 
```

Therefore, git-xargs is able to perform all safe work as quickly as possible by leveraging goroutines, while treating the more expensive open pull request API calls separately. 

 The `--max-concurrent-repos` flag allows you to specify the upper limit of concurrent goroutines that should be allowed to carry out concurrency-safe work. 
 
Pull requests are also handled on a separate channel so that they can be buffered and retried in accordance with what rate limiting feedback git-xargs is receiving from GitHub's API.  

> 
> Meanwhile, git-xargs flows all open pull request jobs through a separate time-buffered channel.

Author: ```suggestion
```

**Distinct processing channels for expensive and non-expensive work**

git-xargs distinguishes between work that is safe to perform in parallel, such as certain git operations, and work that should be done with consideration and care, such as issuing open pull requests to GitHub's API. 

Therefore, git-xargs is able to perform all safe work as quickly as possible by leveraging goroutines, while treating the more expensive open pull request API calls separately. 

 The `--max-concurrent-repos` flag allows you to specify the upper limit of concurrent goroutines that should be allowed to carry out concurrency-safe work. 
 
Pull requests are also handled on a separate channel so that they can be buffered and retried in accordance with what rate limiting feedback git-xargs is receiving from GitHub's API.  

This means that git-xargs performs all the work upfront that it can as quickly as possible, and then moves on to serially process the pull request jobs that have resulted from the upfront work of cloning reposotories, making file changes, checking the git worktree, etc.

> **Automatic pull request retries when rate limited**
> By default, git-xargs will re-attempt opening a pull request that failed due to rate limiting. The `--max-pull-request-retries` flag allows you to specify how many times you'd like a given pull request to be re-attempted in case of failure due to rate limiting. By default, the value is `3`, meaning that if you do not pass this flag, `git-xargs` will retry all rate-limit-blocked pull requests 3 times.

Author: ```suggestion
**Automatic pull request retries when rate limited**

By default, git-xargs will re-attempt opening a pull request that failed due to rate limiting. The `--max-pull-request-retries` flag allows you to specify how many times you'd like a given pull request to be re-attempted in case of failure due to rate limiting. By default, the value is `3`, meaning that if you do not pass this flag, `git-xargs` will retry all rate-limit-blocked pull requests 3 times.
```

When it is provided by GitHub, git-xargs will instead use the values of any `Retry-After` header, or the delta of seconds between the current time and the rate limit error's reset time. When these two values are not available, git-xargs falls back to the user-specified value of `--seconds-to-wait-when-ratelimited`, if the flag was passed, or the default value, which is 60 seconds.

**Specifying the pause between pull requests**
The GitHub's API [integration guidelines](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits) specify that clients should pause at least 1 second in between consecutive requests against expensive endpoints, such as the one for opening a new pull request. As a result, git-xargs defaults to pausing 1 second in between each pull request that is opened. The flag `--seconds-between-prs` allows you to modify this value. For example, if you were to pass `--seconds-between-prs 30`, then git-xargs will sleep for half a minute between issuing pull request API calls to GitHub.

**Guidelines and observations from testing**

> In local testing, the actual thresholds for when GitHub's secondary rate limits kick in can vary depending on the number of repos you're targeting and how long your script or command takes to complete. Secondary rate limits have been observed on jobs targeting as few as 10 repositories. If you are the rate limiting flags with reasonable values, your job should never be rate-limited in the first place. 

Author: ```suggestion
In local testing, the actual thresholds for when GitHub's secondary rate limits kick in can vary depending on the number of repos you're targeting and how long your script or command takes to complete. Secondary rate limits have been observed on jobs targeting as few as 10 repositories. If you are using the rate limiting flags with reasonable values, your job should never be rate-limited in the first place. 
```

```

## Rate Limiting

git-xargs attempts to be a good citizen as regards consumption of the GitHub API. git-xargs conforms to GitHub's API [integration guidelines](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits).

> In addition, git-xargs includes several features and flags that assist in running your jobs in a manner that does not trip GitHub's rate limits:

Reviewer 2: ```suggestion
In addition, git-xargs includes several features and flags to help you: 

1. run jobs without tripping GitHub's rate limits
1. recover when rate limited by automatically re-trying failed pull requests again, still while honoring the GitHub rate limits
```

**Distinct processing channels for expensive and non-expensive work**
git-xargs performs all work up to issuing the API call to open a pull request in parallel - according to the concurrency limits the user might optionally specify by way of the `--max-concurrent-repos` flag.

Meanwhile, git-xargs flows all open pull request jobs through a separate time-buffered channel.

This means that git-xargs performs all the work upfront that it can as quickly as possible, and then moves onto serially process the pull request jobs that have resulted from the upfront work of cloning reposotories, making file changes, checking the git worktree, etc.

**Automatic pull request retries when rate limited**
By default, git-xargs will re-attempt opening a pull request that failed due to rate limiting. The `--max-pull-request-retries` flag allows you to specify how many times you'd like a given pull request to be re-attempted in case of failure due to rate limiting. By default, the value is `3`, meaning that if you do not pass this flag, `git-xargs` will retry all rate-limit-blocked pull requests 3 times.

**Automatic backoff when rate limiting is detected**

> When git-xargs detects that it has been rate limited by GitHub, it begins requeuing failed pull requests for retry, but with an additional, larger buffer of time in between the next attemp. This larger buffer of time is intended to allow GitHub rate limit status to return to normal. The `--seconds-to-wait-when-ratelimited` flag specifies the number of seconds in addition to wait when git-xargs detects it has been rate limited. Note that this extra buffer of time is in addition to the value specified by the `--seconds-between-prs` flag.

Reviewer 2: ```suggestion
When git-xargs detects that it has been rate limited by GitHub, it begins requeuing failed pull requests for retry, but with an additional, larger buffer of time in between the next attempt. This larger buffer of time is intended to allow GitHub rate limit status to return to normal. The `--seconds-to-wait-when-ratelimited` flag specifies the number of seconds in addition to wait when git-xargs detects it has been rate limited. Note that this extra buffer of time is in addition to the value specified by the `--seconds-between-prs` flag.
```

When it is provided by GitHub, git-xargs will instead use the values of any `Retry-After` header, or the delta of seconds between the current time and the rate limit error's reset time. When these two values are not available, git-xargs falls back to the user-specified value of `--seconds-to-wait-when-ratelimited`, if the flag was passed, or the default value, which is 60 seconds.

**Specifying the pause between pull requests**

> The GitHub's API [integration guidelines](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits) specify that clients should pause at least 1 second in between consecutive requests against expensive endpoints, such as the one for opening a new pull request. As a result, git-xargs defaults to pausing 1 second in between each pull request that is opened. The flag `--seconds-between-prs` allows you to modify this value, so, for example, if you were to pass `--seconds-between-prs 30`

Reviewer 2: ```suggestion
The GitHub's API [integration guidelines](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits) specify that clients should pause at least 1 second in between consecutive requests against expensive endpoints, such as the one for opening a new pull request. As a result, git-xargs defaults to pausing 1 second in between each pull request that is opened. The flag `--seconds-between-prs` allows you to modify this value. For example, if you were to pass `--seconds-between-prs 30`, then git-xargs will sleep for half a minute between issuing pull request API calls to GitHub.
```

| `--max-concurrent-repos`              | Limits the number of concurrent processed repositories. This is only useful if you encounter issues and need throttling when running on a very large number of repos. Default is `0` (Unlimited)                                                                                                                                                                                                                              | Integer | No       |
| `--draft`                             | Whether to open pull requests in draft mode. Draft pull requests are available for public GitHub repositories and private repositories in GitHub tiered accounts. See [Draft Pull Requests](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests) for more details.                                                   | Boolean | No       |

> | `--seconds-between-prs`               | The number of seconds to wait between opening serial pull requests. If you are being rate limited, continue to increase this value until rate limiting eases. Note, this value cannot be negative, so if you pass a value less than 1, the seconds to wait between pull requests will be set to 1 second.                                                                                                                     | Integer | No       |

Reviewer: Could you incorporate `Default: 1` here like for `--loglevel`?

Reviewer: NIT: In fact for all the non-required args, it would be nice to say what the default is explicitly.

Author: Agreed! I'll add this in.

Author: Done in [57526e2](https://github.com/gruntwork-io/git-xargs/pull/67/commits/57526e26aa4aa61740fbbba2476fcbfe1535ea92)

| ------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- | -------- |
| `--branch-name`                       | You must specify the name of the branch to make your local and remote changes on. You can further control branching behavior via `--skip-pull-requests` as explained below                                                                                                                                                                                                                                                                      | String  | Yes      |

> | `--loglevel`                          | Specify the log level of messages git-xargs should print to STDOUT at runtime. By default, this is INFO - so only INFO level messages will be visible. Pass DEBUG to see runtime errors encountered by your scripts or commands. Accepted levels are TRACE, DEBUG, INFO, WARNING, ERROR, FATAL and PANIC. Default: `INFO`.                                                                                                                      | String  | No       |

Reviewer: (It makes it easier to visually scan when looking for defaults)

> | `--loglevel`                          | Specify the log level of messages git-xargs should print to STDOUT at runtime. By default, this is INFO - so only INFO level messages will be visible. Pass DEBUG to see runtime errors encountered by your scripts or commands. Accepted levels are TRACE, DEBUG, INFO, WARNING, ERROR, FATAL and PANIC. Default: `INFO`.                                                                                                                      | String  | No       |

Reviewer: One more note: Everything that is not required should say what its default value is, unless there's some reason for not having it displayed.

> | `--repos`                             | If you want to specify many repos and manage them in files (which makes batching and testing easier) then use this flag to pass the filepath to a repos file. See [the repos file format](#option-2-flat-file-of-repository-names) for more information                                                                                                                                                                                         | String  | No       |

Reviewer: NIT: period at the end? Or not. I see a mix of periods and not periods at the ends of things in this section, like lines 415-416 have periods.
